ReactNativeClient/lib/services/searchengine/queryBuilder.ts

```diff
@@ -0,0 +1,361 @@
+		return `child_notebooks(id) as (select folders.id from folders where id IN (select id from folders where ${likes.join(' OR ')})
+		union all select folders.id from folders JOIN child_notebooks on folders.parent_id=child_notebooks.id)`;
+	}
+};
+
+const hyphenateDate = (date: string) => `${date.slice(0,4)}-${date.slice(4, 6)}-${date.slice(6, 8)}`;
+
+const getUnixMs = (date:string) => Date.parse(hyphenateDate(date));
+
+const getNextMonthDate = (value: string) : string => {
+	let endYear = value.slice(0, 4);
+	let endMonth = value.slice(-2);
+	const incYear: boolean = ((Number(endMonth) + 1) / 12) > 1.0;
+	endMonth = ((Number(endMonth) + 1) % 12).toString();
+	if (endMonth.length == 1) endMonth = `0${endMonth}`;
+	if (incYear) endYear = (Number(endYear) + 1).toString();
+	return `${endYear + endMonth}01`;
+};
+
+const makeDateFilter = (type: string, values: string[], queryParts: string[], params: string[], negated: boolean = false, relation: string): void => {
+	const yyyymmdd = /^[0-9]{8}$/;
+	const yyyymm = /^[0-9]{6}$/;
+	const yyyy = /^[0-9]{4}$/;
+	const smartValue = /^(day|week|month|year)-([0-9]+)$/i;
+
+	const msInDay = 60 * 60 * 24 * 1000;
+
+	values.forEach(value => {
+		queryParts.push(`${relation} ROWID IN (SELECT ROWID from notes_normalized where notes_normalized.user_${type}_time ${negated ? '<' : '>='} ?)`);
+		if (yyyymmdd.test(value)) {
+			const msOfDay = getUnixMs(value);
+			const msOfNextDay = msOfDay + msInDay;
+			negated ? params.push(msOfNextDay.toString()) : params.push(msOfDay.toString()) ;
+		} else if (yyyymm.test(value)) {
+			const msOfMonth = getUnixMs(`${value}01`); // make day start from the 1st day of the month
+			const msOfNextMonth = getUnixMs(getNextMonthDate(value));
+			negated ? params.push(msOfNextMonth.toString()) : params.push(msOfMonth.toString());
+		} else if (yyyy.test(value)) {
+			const msOfYear = getUnixMs(`${value}0101`); // make day start from the 1st day of the month 1st month of year
+			const msOfNextYear = getUnixMs(`${(Number(value) + 1).toString()}0101`);
+			negated ? params.push(msOfNextYear.toString()) : params.push(msOfYear.toString());
+		} else if (smartValue.test(value)) {
+			const match = smartValue.exec(value);
+			const timeUnit = match[1]; // eg. day, week, month, year
+			const n = Number(match[2]); // eg. 1, 12, 101
+			const msStart = time.goBackInTime(n, timeUnit); // eg. goBackInTime(1, 'day')
+			let msEnd = null;
+			switch (timeUnit) {
+			case 'day': msEnd = Number(msStart) + msInDay; break;
+			case 'week': msEnd = Number(msStart) + (msInDay * 7); break;
+			case 'month': msEnd = Number(msStart) + (msInDay * 7 * 30); break; // not accurate! maybe use goForwardInTime(startTime?, n, timeUnit)
+			case 'year': msEnd = Number(msStart) + (msInDay * 7 * 30 * 12); break;  // not accurate!
+			}
+			negated ? params.push(msEnd.toString()) : params.push(msStart.toString());
+		} else {
+			throw new Error(`Date value of ${type} in unknown format: ${value}`);
+		}
+	});
+};
```

Comment:
There's a lot of date formatting in here. Would it not be possible to use the `moment` package (which we already import) for this?